File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/피로도/Solution.java
class Solution {

    static int maximum = 0;

    public int solution(int k, int[][] dungeons) {
        backTracking(k, dungeons, new boolean[dungeons.length], 0);
        return maximum;
    }

    private void backTracking(int k, int[][] dungeons, boolean[] visited, int count) {
        if (count > maximum) {
            maximum = count;
        }

        for (int i = 0; i < dungeons.length; i++) {
            if (visited[i] || k < dungeons[i][0]) {
                continue;
            }

            visited[i] = true;
            backTracking(k - dungeons[i][1], dungeons, visited, count + 1);
            visited[i] = false;
        }
    }
}

File: 킬링캠프/1주 차 - 완전탐색, 백트래킹, 이분탐색/challenge/피로도/solution.py
def solution(k, dungeons):
    # 전역 값
    answer = 0
    
    # 메서드
    def search(k, count, visited):
        nonlocal answer
        answer = max(answer, count)

        for i in range(len(dungeons)):
            needHp, useHp = dungeons[i]
            if visited[i] or k < needHp:
                continue

            visited[i] = True
            search(k - useHp, count + 1, visited)
            visited[i] = False
    
    # 메인 로직
    visited = [False for _ in dungeons]
    search(k, 0, visited)

    return answer
